concent_api/middleman/middleman_server.py

```
@@ -0,0 +1,94 @@
+        self._loop.add_signal_handler(signal.SIGTERM, self._terminate_connection)
+
+    def run(self) -> None:
+        try:
+            # start MiddleMan server
+            logger.info("Starting MiddleMan")
+            self._start_middleman()
+        except OSError as exception:
+            logger.error(
+                f"Exception <OSError> occurred while starting MiddleMan server for Concent: {str(exception)}")
+            sys.exit(EXIT_STATUS_ADDRESS_IN_USE_ERROR)
+
+        try:
+            # Serve requests until Ctrl+C is pressed
+            logger.info('MiddleMan is serving on {}'.format(self._server_for_concent.sockets[0].getsockname()))
+            self._run_forever()
+        except KeyboardInterrupt:
+            logger.info("Ctrl-C has been pressed.")
+
+        # Close the server
+        logger.info("Server is closing...")
+        self._close_middleman()
+        logger.info("Closed.")
+        exit(0)
+
+    def _run_forever(self):
+        self._loop.run_forever()
+
+    def _start_middleman(self):
+        server_coroutine = asyncio.start_server(
+            self._handle_user_request,
+            self._bind_address, self._port,
+            loop=self._loop
+        )
+        self._server_for_concent = self._loop.run_until_complete(server_coroutine)
+
+    def _close_middleman(self):
+        self._server_for_concent.close()
+        self._loop.run_until_complete(self._server_for_concent.wait_closed())
+        self._loop.close()
+
+    @safe_execution(logger, crash_logger)
+    async def _handle_user_request(self, reader, writer) -> None:
+        data = await reader.readuntil()
+        remote_address = writer.get_extra_info('peername')
+        print("Received %r from %r" % (data, remote_address))
+        self._connection_counter += 1
+        await self._respond_to_user(data, writer)
+
+    @safe_execution(logger, crash_logger)
+    async def _respond_to_user(self, data: bytes, writer: asyncio.StreamWriter) -> None:
```

Same as above. Can't you just let the coroutine raise exceptions and handle the futures?